docs: glossary for keyboard author (#1827)
keymanapp/help.keyman.com: developer/language/guide/glossary.md

---
title: Glossary for Keyboard Author 
---

## B {#b-glossary}

### BCP 47 language tag
A standardized code that is used to identify human languages on the Internet. More on [BCP 47](https://www.rfc-editor.org/info/bcp47).

There are many possible subtags, but only three types are currently used in most places in Keyman Developer:
* [language subtag](../../current-version/reference/bcp-47#toc-the-language-subtag)
* [script subtag](../../current-version/reference/bcp-47#toc-the-script-subtag)
* [region subtag](../../current-version/reference/bcp-47#toc-the-region-subtag)

More information about BCP 47 language tag and full details on [how they are used in Keyman Developer](../../current-version/reference/bcp-47).

## D {#d-glossary}

> ### Deadkey (Or Dead Key, shorthand dk)

Reviewer: ```suggestion
### Deadkey (Or Dead Key, shorthand: dk)
```

* [KVKS](../../current-version/reference/file-types/kvks): a Keyman On Screen Keyboard source file.
* [TSV](../../current-version/reference/file-types/tsv): or a tab-separated values file contains a word list. This word list is used to predict and correct words using the predictive text functionality.
* [TTF](../../current-version/reference/file-types/ttf): a TrueType font file. This file is distributed with a keyboard package to be used as a keyboard font or display font.
* [MODEL.JS](../../current-version/reference/file-types/model-js): compiled in Keyman Developer from the Keyman lexical model source (.MODEL.TS) file and (.TSV) wordlist.
* [MODEL.TS](../../current-version/reference/file-types/model-ts): a [lexical model](#toc-lexical-model) definition source file.
* [More File types](../../current-version/reference/file-types/).

### File layout
Your Keyman keyboard is grouped into one of the three folders:
* release: keyboards with full source that pass our stringent quality checks, Unicode only, generally no PUA
* legacy: binary distributions of keyboards, migrated from Tavultesoft servers, Unicode and non-Unicode

> * experimental: keyboards that have not yet passed the quality check, available from the 'experimental' category in Keyman apps with Keyman 10 and later. Unicode only, but PUA is acceptable.

Reviewer: This doesn't make much sense to me here. Does the user see these folders unless they clone the repo? Might be better to explain the file layout for a keyboard project.

Visit [Keyman Developer structure for standard keyboard source file folders](../../current-version/reference/file-layout) for more information.

### Font
A font is a graphical representation of text that may include a different typeface, point size, weight, color, or design. If you've read or received a message via Keyman that is displayed as square boxes, the most likely cause is that your device does not support the language or script. This is how to properly [include font(s) into the keyboard package](../../keyboards/github/step-2#toc-including-fonts-in-your-keyboard-package), and [how to fix square boxes related to font](https://keyman.com/fonts/).

## G {#g-glossary}

### Gesture
A gesture is a finger motion like long press, flick, and multitap that can be used in Keyman.

## K {#k-glossary}

### Key Code (Or Virtual Key Code)
Keyman uses the same keycode across all operating system. More on [Virtual Key Code](./virtual-keys#using-virtual-keys).

### Keyboard

> Keyboard may also mean:

Reviewer: ```suggestion
Keyboard may mean:
```

* Physical Keyboard - the hardware keyboard used for text input with a computer.
* On-screen Keyboard - or OSK, a keyboard displayed graphically which represents the layout of the keys on the physical keyboard and what characters they will produce.
* Touch Keyboard - the keyboard displayed for text input on a mobile device with a touchscreen such as a smartphone or tablet.

### Keyboard Layout
Any specific physical, visual, or functional arrangement of the keys of a computer. In general every key on a keyboard can be identified by either its position on the keyboard or the character it would print out. In Keyman a key can be referenced by its position on the US English Keyboard ( K_A, K_B, etc.). In this case a [positional keyboard layout](#toc-positional-keyboard-layout) is used. A key can also be referenced by its output ( “A” , “B”, etc.). In this case a [mnemonic keyboard layout](#toc-mnemonic-keyboard-layout) is used.

### Keyboard Repository

> Keyman repository is grouped into one of the three folders:

Reviewer: ```suggestion
In the Keyman [keyboards repository](https://github.com/keymanapp/keyboards) keyboards are grouped into three folders:
```

> * release: keyboards with full source that pass our stringent quality checks, Unicode only, generally no PUA

Author: I can add the link to this: https://en.wikipedia.org/wiki/Private_Use_Areas.

* legacy: binary distributions of keyboards, migrated from Tavultesoft servers, Unicode and non-Unicode
* experimental: keyboards that have not yet passed the quality check, available from the 'experimental' category in Keyman apps with Keyman 10 and later. Unicode only, but PUA is acceptable.

[Find out how to distribute your keyboard and work with Keyman Keyboard Repository](../../keyboards/).

## L {#l-glossary}

### Language Structure
The Keyman keyboard programming language is a language designed around transforming inputs to outputs, through rules rather than through static tables. Please refer to the [overview of the Keyman language](overview).

### Lexical Model
A wordlist and a set of rules of how the word is interpreted. The way your keyboard knows how to suggest predictions and corrections for your language is through its lexical model. [Visit the Lexical Model page](../../current-version/guides/lexical-models/intro/#toc-what-is-a-lexical-model-).

## M {#m-glossary}

### Mnemonic Keyboard Layout

> In a mnemonic keyboard layout keys are identified by the character it prints out. For example, the key that prints out the character **'z'**:

Reviewer: I think it would be good to mention and cross-reference positional keyboard layouts in this entry.

* On a US English Keyboard, **'z'** is printed out by pressing K_Z (labeled 'Z' : the leftmost character key on the bottom row)
* On a German Keyboard, **'z'** is printed out by pressing K_Y (labeled 'Z' : a key somewhat in the middle of the first row)
* On a French Keyboard, **'z'** is printed out by pressing K_W (labeled 'Z' : the second key to the left of the first row)

> If we, for example, specified a Keyboard rule in Keyman Developer for a US layout (`+ K_Z  >  'z'`) and used that mnemonic layout with a German physical keyboard for key K_Y (where the keycap is labeled 'Z'), still a **'z'** will be printed out.

Reviewer: It might be better to start from the output, e.g. something like

```suggestion
If we, for example, specified a Keyboard rule in Keyman Developer for a US layout (`+ K_Z  >  'z'`) and used that mnemonic layout with a German physical keyboard, then to print out **'z'** the key K_Y (where the keycap is labeled 'Z') will have to be pressed.
```

This means when we specify a key in a mnemonic layout we talk about a key that produces a certain output (e.g. 'z') no matter where it is positioned at in different keyboard layouts.

### Modifier Keys
Keys on a Computer Keyboard like <kbd>Shift</kbd>, <kbd>Ctrl</kbd>, <kbd>Alt</kbd>, <kbd>AltGr</kbd>, <kbd>Option</kbd> etc. Modifier Keys may be pressed simultaneously with certain other keys to activate predefined functions. They may also be used to produce different characters. For example, on a German keyboard pressing the Q-Key produces 'q', Shift + Q-Key produces 'Q' whereas AltGr + Q-Key produces '@'. [More on Modifier Key](../../core/current-version/#toc-glossary).

## P {#p-glossary}

### Positional Keyboard Layout

> In a positional keyboard layout keys of a keyboard are identified by the position on the keyboard. For example, the leftmost character key on the bottom row is always called **K_Z**:

Reviewer: see above - cross-reference mnemonic keyboard layout.

* On a US English Keyboard, the key is called **K_Z**, the keycap is labeled 'Z' and it prints out 'z'
* On a German Keyboard, the key is called **K_Z**, the keycap is labeled 'Y' and it prints out 'y'
* On a French Keyboard the key is called **K_Z**, the keycap is labeled 'W' and it prints out 'w'

> If we, for example, specified a Keyboard rule in Keyman Developer for a US layout (`+ K_Z  >  'z'`) and used that positional layout with a German physical keyboard for key K_Z (where the keycap is labeled 'Y') will print out **'z'**.

Reviewer: See above

```suggestion
If we, for example, specified a Keyboard rule in Keyman Developer for a US layout (`+ K_Z  >  'z'`) and used that positional layout with a German physical keyboard, then to print out **'z'** the key K_Z (where the keycap is labeled 'Y') will have to be pressed.
```

│   keyboard.kpj.user: A user preference file. You can safely delete it at any time.
│
├───build
|       keyboard.keyboard_info: A metadata file detailing the keyboard's origin, version, requirements, and capabilities.
│       keyboard.kmp:           The installable package file.
│       keyboard.kmx:           The compiled keyboard. This intermediate file should not be distributed.
│       keyboard.kvk:           The compiled on screen keyboard. This intermediate file should not be distributed.
│       keyboard.js:            The keyboard compiled to Javascript for use with KeymanWeb.
│
└───source
        keyboard.ico:                 The icon of the keyboard.
        readme.htm:                   Introductory web page for end users of your keyboard.
        welcome.htm:                  A web page which describes how to use your keyboard.
        keyboard.keyman-touch-layout: A touch layout file description.
        keyboard.kmn:                 The keyboard source file.
        keyboard.kps:                 The package source, which is used to create a compressed .kmp.
        keyboard.kvks:                The on screen keyboard template for desktop platforms.

> ```

Reviewer 2: I am not sure about the advisability of including lots of content in the glossary that is also elsewhere. It's likely to get out of sync in the future as we edit. (@sgschantz you like glossaries, what do you think?)

Reviewer 2: Tagging @sgschantz again

Reviewer 3: It can be painful to maintain two locations with overlapping content because they are eventually going to diverge. 

If we look at it from the perspective of the users, then we need to ask what will make their lives easier. I think it is justified to have two glossaries if the target audience is different enough. We also don't want either audience to have to know about both glossaries and check the second if the first does not contain the term they need. It's hard to remember where one is located.

So that means that we either repeat information to make it easy for every that's using either glossary, or the glossaries link to each other. Linking between glossaries is still pretty messy, because the user is still jumping back and forth.

I guess repeating information is probably best while possibly simplifying the entry if it is for a less technical audience. 

The only other option I can think of is to have a single glossary that contains everything and is not specialized for different types of users. Maybe it could still cater to multiple audiences with an indicator on some entries as to identify them as more technical.

Author: Thank you for your suggestions @sgschantz. 

How about we just have this to show file and folder structure, remove the short description, and with the use of a link (https://help.keyman.com/developer/current-version/reference/file-layout) to allow user to read more?

This way; I'm taking your concern into consideration, @mcdurdin; the future changes won't be too out of sync or hard to update.

Therefore, this glossary will provide just enough information for a less technical audience while also be a gateway to a more technical information.

Hopefully I have concluded it in the middle, or have I not made any progress on the concern?

Reviewer 2: @sgschantz, this wasn't really about two glossaries, but rather the glossary copying content out of other documentation. On multiple glossaries, in terms of two different audiences, I think that will necessitate duplication. The glossaries we have for Keyman team developers will overlap with info for integrators and also with info for keyboard authors. I'm okay with that. Ideally the content for each term is tailored to the target audience, and even could be cross-linked between glossaries. 

@Meng-Heng wrote:
> How about we just have this to show file and folder structure, remove the short description, and with the use of a link (https://help.keyman.com/developer/current-version/reference/file-layout) to allow user to read more?

I think rather than showing the file layout here at all, just link to the file layout topic. Just have a single sentence stating that we have a standard file layout for Keyman keyboard projects.

(The file types list may be okay, perhaps take .ttf and .ico out of the list as they are not really specific to Keyman)

> Therefore, this glossary will provide just enough information for a less technical audience while also be a gateway to a more technical information.

@Meng-Heng, that's perfect. A glossary should be a gateway to deeper information.

Author: Thank you, I will make the changes later this afternoon.

* release: keyboards with full source that pass our stringent quality checks, [Unicode](https://en.wikipedia.org/wiki/Unicode) only, generally no [PUA](https://en.wikipedia.org/wiki/Private_Use_Areas)
* legacy: binary distributions of keyboards, migrated from Tavultesoft servers, Unicode and non-Unicode
* experimental: keyboards that have not yet passed the quality check, available from the 'experimental' category in Keyman apps with Keyman 10 and later. Unicode only, but PUA is acceptable.

[Find out how to distribute your keyboard and work with Keyman Keyboard Repository](../../keyboards/).

## L {#l-glossary}

### Language Structure
The Keyman keyboard programming language is a language designed around transforming inputs to outputs, through rules rather than through static tables. Please refer to the [overview of the Keyman language](overview).

### Lexical Model
A wordlist and a set of rules of how the word is interpreted. The way your keyboard knows how to suggest predictions and corrections for your language is through its lexical model. [Visit the Lexical Model page](../../current-version/guides/lexical-models/intro/#toc-what-is-a-lexical-model-).

## M {#m-glossary}

> ### Mnemonic Keyboard Layout

Reviewer 2: We should link to the &mnemoniclayout system store documentation here too

## P {#p-glossary}

### Positional Keyboard Layout [`store(&mnemoniclayout)`](../reference/mnemoniclayout)
In a positional keyboard layout, contrary to a [mnemonic keyboard layout](#toc-mnemonic-keyboard-layout), keys are identified by the position on the keyboard. For example, the leftmost character key on the bottom row is always called **K_Z**:

* On a US English Keyboard, the key is called **K_Z**, the keycap is labeled 'Z' and it prints out 'z'
* On a German Keyboard, the key is called **K_Z**, the keycap is labeled 'Y' and it prints out 'y'
* On a French Keyboard the key is called **K_Z**, the keycap is labeled 'W' and it prints out 'w'

If we, for example, specified a Keyboard rule in Keyman Developer for a US layout (`+ K_Z  >  'z'`) and used that positional layout with a German physical keyboard, then to print out **'z'** the key K_Z (where the keycap is labeled 'Y') will have to be pressed.

> This means when we specify a key in a positional layout we talk about a certain position (e.g. K_Z) no matter may be written on a keycap for a specific keyboard layout.

Reviewer: ```suggestion
This means when we specify a key in a positional layout we talk about a certain position (e.g. K_Z) no matter what is written on a keycap for a specific keyboard layout.
```
